cocotb/triggers.py

```diff
@@ -915,9 +918,21 @@ async def with_timeout(trigger, timeout_time, timeout_unit="step"):
 
     .. versionadded:: 1.3
 
+    .. versionchanged:: 1.6
+        Support passing :term:`python:coroutine`\ s.
+
     .. deprecated:: 1.5
         Using None as the the *timeout_unit* argument is deprecated, use ``'step'`` instead.
-   """
+    """
+
+    if inspect.iscoroutine(trigger):
+        task = cocotb.fork(trigger)
+        try:
+            res = await with_timeout(task, timeout_time, timeout_unit)
+        except cocotb.result.SimTimeoutError:
+            task.kill()
+            raise
+        return res
```

Comment:
If the user passes `timeout_unit is None` and has `warnings.simplefilter('error')`, then this throws an exception but does not kill the task.

Reply:
Or if the task throws an exception before the first `await` due to how `fork` works. Basically the issue is that it doesn't kill the coroutine in the case of *any* exception, `SimTimeoutError` could be changed to `Exception`.

Reply:
In the case I describe, the coroutine should not even have been started.

Reply:
yeah, I guess we should put task.kill in finally

Reply:
@eric-wieser thoughts?
```diff
diff --git a/cocotb/triggers.py b/cocotb/triggers.py
index 5d4b9d4c..5330a68f 100644
--- a/cocotb/triggers.py
+++ b/cocotb/triggers.py
@@ -948,23 +948,20 @@ async def with_timeout(trigger, timeout_time, timeout_unit="step"):
     .. versionchanged:: 1.6
         Support passing :term:`python:coroutine`\ s.
     """
-
-    if inspect.iscoroutine(trigger):
-        task = cocotb.fork(trigger)
-        try:
-            res = await with_timeout(task, timeout_time, timeout_unit)
-        except cocotb.result.SimTimeoutError:
-            task.kill()
-            raise
-        return res
-
     if timeout_unit is None:
         warnings.warn(
             'Using timeout_unit=None is deprecated, use timeout_unit="step" instead.',
             DeprecationWarning, stacklevel=2)
-        timeout_unit="step"  # don't propagate deprecated value
+        timeout_unit = "step"  # don't propagate deprecated value
+    if inspect.iscoroutine(trigger):
+        shielded = False
+        trigger = cocotb.fork(trigger)
+    else:
+        shielded = True
     timeout_timer = cocotb.triggers.Timer(timeout_time, timeout_unit)
     res = await First(timeout_timer, trigger)
+    if isinstance(trigger, cocotb.decorators.RunningTask) and not shielded:
+        trigger.kill()
     if res is timeout_timer:
         raise cocotb.result.SimTimeoutError
     else:
```